### ISSUE TITLE
Writing to an INT memory, overwrites the memories below it

### ISSUE THREAD

**Issue by ildomarcarvalho** [gijzelaerr/python-snap7#94]
Dear all,
When I try to write to an integer memory, it overwrites some memories below this int memory. Can you help me to find out if it is a bug or I just using the the snap7 incorrectly?

```python
import snap7

layout = """
      0       DB_VAR_1                    INT
      2       DB_VAR_2                    INT
      4.0     IN_Reserva_Camera           BOOL
      4.1     IN_Sinal_Iluminacao         BOOL
      4.2     IN_Robo_Trigger_Cam         BOOL
      4.3     IN_Robo_UR5_Inspecao        BOOL
      4.4     IN_Robo_UR5_Operacao        BOOL
      4.5     IN_Robo_UR5_Sinal_4         BOOL
      4.6     IN_Robo_UR5_Iniciar_ACK     BOOL
      4.7     IN_Reserva_2                BOOL
      5.0     IN_Sensor_Pressostato       BOOL
      5.1     IN_Reserva_Sensor_1         BOOL
      5.2     IN_Mon_BT_Emer_Painel_Pr    BOOL
      5.3     IN_Mon_BT_Emer_Reserva_1    BOOL
      5.4     IN_Mon_BT_Emer_Reserva_2    BOOL
      5.5     IN_Mon_Rele_BT_Emer         BOOL
      5.6     IN_Mon_Rele_Chave_Porta     BOOL
      5.7     IN_Reserva_3                BOOL
      6.0     IN_Sinal_MGB_Pos_Porta      BOOL
      6.1     IN_Sinal_MGB_Pos_Bolt       BOOL
      6.2     IN_Sinal_MGB_Pos_Trava      BOOL
      6.3     IN_Sinal_MGB_Diag_Saida     BOOL
      6.4     IN_Solic_Abert_Porta_MGB    BOOL
      6.5     IN_Sinal_Reset_MGB          BOOL
      6.6     IN_Reserva_4                BOOL
      6.7     IN_Reserva_5                BOOL
      7.0     IN_BT_Lig_Deslig_Comando    BOOL
      7.1     IN_BT_Pos_Inicial           BOOL
      7.2     IN_BT_Reset_Falha           BOOL
      7.3     IN_Cha_Manual_Automatico    BOOL
      7.4     IN_Reserva_6                BOOL
      7.5     IN_Reserva_7                BOOL
      7.6     IN_Reserva_8                BOOL
      7.7     IN_Reserva_9                BOOL
      8.0     IN_M_D_Ent_Trafo_Ant_Sec    BOOL
      8.1     IN_M_D_Sai_Trafo_Ant_Sec    BOOL
      8.2     IN_M_D_Ent_Med_Energia      BOOL
      8.3     IN_M_D_Vent_Painel          BOOL
      8.4     IN_M_D_Fonte_POE_Cam        BOOL
      8.5     IN_M_D_Controlador_UR5      BOOL
      8.6     IN_M_D_Luminaria_Painel     BOOL
      8.7     IN_M_D_Medidor_Energia      BOOL
      9.0     IN_M_D_Tomada_Manut         BOOL
      9.1     IN_M_D_Comando_Seguranca    BOOL
      9.2     IN_Reserva_10               BOOL
      9.3     IN_Reserva_11               BOOL
      9.4     IN_Reserva_12               BOOL
      9.5     IN_Reserva_13               BOOL
      9.6     IN_Reserva_14               BOOL
      9.7     IN_Reserva_15               BOOL
      10.0    IN_Sens_Cil_Index_Avanc     BOOL
      10.1    IN_Sens_Presenca_Prod_1     BOOL
      10.2    IN_Sens_Setup_2             BOOL
      10.3    IN_Sens_Setup_3             BOOL
      10.4    IN_Sens_Distrib_Res_1       BOOL
      10.5    IN_Sens_Distrib_Res_2       BOOL
      10.6    IN_Sens_Distrib_Res_3       BOOL
      10.7    IN_Sens_Distrib_Res_4       BOOL
      11.0    IN_Sens_Cil_Index_Recuad    BOOL
      11.1    IN_Sens_Presenca_Prod_2     BOOL
      11.2    IN_Sens_Setup_1             BOOL
      11.3    IN_Sens_Setup_4             BOOL
      11.4    IN_Sens_Distrib_Res_5       BOOL
      11.5    IN_Sens_Distrib_Res_6       BOOL
      11.6    IN_Sens_Distrib_Res_7       BOOL
      11.7    IN_Sens_Distrib_Res_8       BOOL
      12.0    OUT_Elem_Incolor_Torre      BOOL
      12.1    OUT_Elem_Azul_Torre         BOOL
      12.2    OUT_Elem_Amarelo_Torre      BOOL
      12.3    OUT_Elem_Vermelho_Torre     BOOL
      12.4    OUT_Elem_Verde_Torre        BOOL
      12.5    OUT_Trigger_Camera          BOOL
      12.6    OUT_Restart_1_UR5           BOOL
      12.7    OUT_Restart_2_UR5           BOOL
      13.0    OUT_Robo_UR5_Inic_Ciclo     BOOL
      13.1    OUT_Robo_UR5_Sinal_2        BOOL
      13.2    OUT_Robo_UR5_Sinal_3        BOOL
      13.3    OUT_Robo_UR5_Sinal_4        BOOL
      13.4    OUT_Robo_Setup_1            BOOL
      13.5    OUT_Robo_Setup_2            BOOL
      13.6    OUT_LigaI_Iluminacao        BOOL
      13.7    OUT_Reserva_4               BOOL
      14.0    OUT_Valv_Avan_Cil_Index     BOOL
      14.1    OUT_Valv_Rec_Cil_Indexad    BOOL
      14.2    OUT_Valv_Reserva_1          BOOL
      14.3    OUT_Valv_Reserva_2          BOOL
      14.4    OUT_Valv_Reserva_3          BOOL
      14.5    OUT_Valv_Reserva_4          BOOL
      14.6    OUT_Lib_T_Chave_PortA       BOOL
      14.7    OUT_Reserva_5               BOOL
      15.0    OUT_Sinal_BT_Lig_Des_Com    BOOL
      15.1    OUT_Sinal_BT_Pos_Inicial    BOOL
      15.2    OUT_Sinal_BT_Reset_Falha    BOOL
      15.3    OUT_Reserva_6               BOOL
      15.4    OUT_Reserva_7               BOOL
      15.5    OUT_Sinal_MGB_BT_Reset      BOOL
      15.6    OUT_Sin_MGB_BT_Solic_Abe    BOOL
      15.7    OUT_Reserva_8               BOOL
      16.0    PLC_Setup_1                 BOOL
      16.1    PLC_Setup_2                 BOOL
      16.2    PLC_Start_Cycle             BOOL
      16.3    PLC_Inspecao_Finalizada     BOOL
      16.4    PLC_ACK                     BOOL
      18      PLC_Cod_Rastreabilidade     STRING[254]
      274     PLC_Tempo_Total             REAL
      278.0   IHM_Start_Cycle_Confirm     BOOL
      280     IHM_Numero_Insp_1           INT
      282.0   IHM_Resultado_Insp_1        BOOL
      284     IHM_Gap_1                   REAL
      288     IHM_Numero_Insp_2           INT
      290.0   IHM_Resultado_Insp_2        BOOL
      292     IHM_Gap_2                   REAL
      296     IHM_Numero_Insp_3           INT
      298.0   IHM_Resultado_Insp_3        BOOL
      300     IHM_Gap_3                   REAL
      304     IHM_Numero_Insp_4           INT
      306.0   IHM_Resultado_Insp_4        BOOL
      308     IHM_Gap_4                   REAL
      312     IHM_Numero_Insp_5           INT
      314.0   IHM_Resultado_Insp_5        BOOL
      316     IHM_Gap_5                   REAL
      320     IHM_Numero_Insp_6           INT
      322.0   IHM_Resultado_Insp_6        BOOL
      324     IHM_Gap_6                   REAL
      328     IHM_Numero_Insp_7           INT
      330.0   IHM_Resultado_Insp_7        BOOL
      332     IHM_Gap_7                   REAL
      336     IHM_Numero_Insp_8           INT
      338.0   IHM_Resultado_Insp_8        BOOL
      340     IHM_Gap_8                   REAL
      344     IHM_Numero_Insp_9           INT
      346.0   IHM_Resultado_Insp_9        BOOL
      348     IHM_Gap_9                   REAL
      352     IHM_Numero_Insp_10          INT
      354.0   IHM_Resultado_Insp_10       BOOL
      356     IHM_Gap_10                  REAL
      360     IHM_Numero_Insp_11          INT
      362.0   IHM_Resultado_Insp_11       BOOL
      364     IHM_Gap_11                  REAL
      368     IHM_Numero_Insp_12          INT
      370.0   IHM_Resultado_Insp_12       BOOL
      372     IHM_Gap_12                  REAL
      376     IHM_Numero_Insp_13          INT
      378.0   IHM_Resultado_Insp_13       BOOL
      380     IHM_Gap_13                  REAL
      384     IHM_Numero_Insp_14          INT
      386.0   IHM_Resultado_Insp_14       BOOL
      388     IHM_Gap_14                  REAL
      392     IHM_Numero_Insp_15          INT
      394.0   IHM_Resultado_Insp_15       BOOL
      396     IHM_Gap_15                  REAL
      400     IHM_Numero_Insp_16          INT
      402.0   IHM_Resultado_Insp_16       BOOL
      404     IHM_Gap_16                  REAL
      408     IHM_Numero_Insp_17          INT
      410.0   IHM_Resultado_Insp_17       BOOL
      412     IHM_Gap_17                  REAL
      416     IHM_Numero_Insp_18          INT
      418.0   IHM_Resultado_Insp_18       BOOL
      420     IHM_Gap_18                  REAL
      424     IHM_Numero_Insp_19          INT
      426.0   IHM_Resultado_Insp_19       BOOL
      428     IHM_Gap_19                  REAL
      432     IHM_Numero_Insp_20          INT
      434.0   IHM_Resultado_Insp_20       BOOL
      436     IHM_Gap_20                  REAL
      440.0   IHM_Resultado_Final         BOOL
      440.1   IHM_Confirm_Enviar          BOOL
      442     IHM_Tempo_Processo          REAL
      446     IHM_Tempo_Estacao           REAL
      450.0   IHM_Keep_Alive_1            BOOL
      450.1   IHM_Keep_Alive_2            BOOL
      450.2   IHM_Salvar                  BOOL
      450.3   IHM_Reinspecionar           BOOL
      452     IHM_Codigo_Erro             INT
      454     IHM_Codigo_Descricao        STRING[254]
      710.0   IHM_Insp_Finalizada_Conf    BOOL
      712     IHM_Passo                   INT
      714     IHM_Passo_2                 INT
      716     IHM_Reserva_1               INT
      718     IHM_Reserva_2               INT
      720     IHM_Reserva_3               INT
      722     IHM_Reserva_4               INT
      724     IHM_Reserva_5               INT
      726.0   FALHA_Pressostato           BOOL
      726.1   FALHA_Presenca_de_Produt    BOOL
      726.2   FALHA_Robo_Home             BOOL
      726.3   FALHA_Distribuidor          BOOL
      726.4   FALHA_Rele_Chave            BOOL
      726.5   FALHA_Rele_BT_Emergencia    BOOL
      726.6   FALHA_Sensor_Setup          BOOL
      726.7   FALHA_Robo_Iniciar_Insp     BOOL
      727.0   FALHA_Robo                  BOOL
      727.1   FALHA_Robo_Emergencia       BOOL
      727.2   FALHA_Robo_Operacao         BOOL
      727.3   FALHA_Cilindro_Index        BOOL
      727.4   FALHA_D_Entr_Trafo_Antes    BOOL
      727.5   FALHA_D_Said_Trafo_Antes    BOOL
      727.6   FALHA_D_Entr_Med_Energia    BOOL
      727.7   FALHA_D_Ventilador_P        BOOL
      728.0   FALHA_D_Fonte_POE           BOOL
      728.1   FALHA_D_Contr_UR5           BOOL
      728.2   FALHA_D_Luminaria_Painel    BOOL
      728.3   FALHA_D_Medidor_Energia     BOOL
      728.4   FALHA_D_Tomada_Manut        BOOL
      728.5   FALHA_D_Fonte_CMD_Seg       BOOL
      728.6   FALHA_IHM                   BOOL
      728.7   FALHA_Maquina               BOOL"""

db_number = 16

plc = snap7.client.Client()
plc.connect('10.0.0.101', 0, 2)

all_data = plc.upload(db_number)

db_16 = snap7.util.DB(
    db_number,
    all_data,
    layout,
    row_size=728+1,
    size=1,
    id_field=None,
    layout_offset=0,
    db_offset=0)

# This piece of code works without any problem
db_16[0].read(plc)
db_16[0]['IHM_Resultado_Insp_1'] = True
db_16[0].write(plc)

# If I try to write to this int memory, I end up writing to the memories
# IHM_Resultado_Insp_1, IHM_Gap_1 and IHM_Numero_Insp_2. Depending of the
# value, even more memories get overwriten (the IHM_Numero_Insp_2 memory gets
# the IHM_Numero_Insp_1 old value.
# The memory IHM_Numero_Insp_1 gets the right value
db_16[0].read(plc)
db_16[0]['IHM_Numero_Insp_1'] = 10
db_16[0].write(plc)

# I made this workaround to write correctly to int memories
def to_bytes(n, length, endianess='big'):
    h = '%x' % n
    s = ('0'*(len(h) % 2) + h).zfill(length*2).decode('hex')
    return s if endianess == 'big' else s[::-1]


IHM_Numero_Insp_1 = int(db_16[0]._specification['IHM_Numero_Insp_1'][0])
plc.db_write(db_number, IHM_Numero_Insp_1, to_bytes(0, 2))
```

Best Regards,

**Comment by spreeker** [comment]
This could be a bug.

This code was used to read out 1000s of sensors and valves and to write
Boolean values. (On/off). Writing int values.might not have been properly
tested!!

There is no code dealing with the endian Ness (I did not need it at the
time)
 and now I see the 'to_int' function has no size check on it, I think its
current implementation is bad It just adds data to the the array and moving
everything if what is inserted is bigger than 2 bytes..leading to very
weird results! (Just tested it in a Python shell) a[2:4] = arraylagerthan2
just 'works' it just inserts everything starting at 2. Oops.

https://github.com/gijzelaerr/python-snap7/blob/master/snap7/util.py#L131

It should be easy for you to fix the offending function 'to_int' and do a
pull request.!

I do not have a PLC lying around anymore to test it properly.

Cheers!

Stephan.

**Comment by ildomarcarvalho** [comment]
Thanks for your reply @spreeker 

I'm sorry I took this long to read your reply, but I posted this just the day I went on vacation.
Unfortunately I don't have a PLC with me anymore too. Maybe in a few weeks I'll have it again, then I will try to fix. If someone reading this post have a PLC, please feel free to fix :)

Thanks!

**Comment by ildomarcarvalho** [comment]
I got my hands in a PLC today, so I could fix and test this bug. I just opened a pull request.

Thanks for the help

**Comment by gijzelaerr** [comment]
I think this issue is solved by accepting #96. Please reopen if you think otherwise :)
